app/models/user.rb

```diff
@@ -122,6 +122,30 @@ def is_admin
     is_global_admin || community_user&.is_admin || false
   end
 
+  def has_ability_on(community_id, ability_internal_id)
+    if is_admin || is_global_moderator
+      return true
+    end
+
+    ability = Ability.where(community_id: community_id).find_by internal_id: ability_internal_id
+
+    if ability.nil?
+      return false
+    end
+
+    cu = community_users.find { |cuf| cuf.community_id == community_id && cuf.id == id }
+
+    if cu.nil?
+      return false
+    end
+
+    if cu.is_moderator || cu.privilege('mod')
+      return true
+    end
+
+    UserAbility.where(community_user_id: cu.id, ability_id: ability.id, is_suspended: false).exists?
```

Comment:
At a first glance, this method feels heavy to me - Rails is usually good at doing this kind of thing in a couple of lines.

We want this method to return true if (correct me if I'm wrong):

 * user is a global mod or admin
 * user is a community mod or admin
 * user has specified ability on specified community

I'd probably go about this by grabbing the community user first, then you can probably do the rest as one or two conditionals. Something like:

```ruby
def has_ability_on(community_id, ability_internal_id)
  # where instead of find so that we're using the database rather than Ruby
  cu = community_users.joins(:community).where(communities: { id: community_id }).first
  if is_admin || is_global_moderator || cu.is_moderator # can't remember the names of all these checks - will need checking!
    true
  else
    UserAbility.joins(:ability).where(community_user_id: cu.id, is_suspended: false, ability: { internal_id: ability_internal_id }).exists?
  end
end
```

(Fair warning: I haven't tested any of that, so if you do change it it'll need testing!)

Reply:
Updated and tested with the following user types: none (logged out), regular (not editor), editor, moderator, admin.  All behaved as expected.

What I have not been able to test in a dev environment is multiple communities.  I don't know how to set that up when everything runs off of localhost not DNS.  I assume some port-wrangling would do it, but maybe we could test this on the dev server?